src/System.Net.Security/src/System/Net/Security/SslStreamPal.Unix.cs

```diff
@@ -160,6 +160,17 @@ public static byte[] ConvertAlpnProtocolListToByteArray(List<SslApplicationProto
                     done = Interop.OpenSsl.DoSslHandshake(((SafeDeleteSslContext)context).SslContext, inputBuffer.token, inputBuffer.offset, inputBuffer.size, out output, out outputSize);
                 }
 
+                // When the handshake is done, and the context is server, check if the alpnHandle was freed during ALPN.
+                // If it was freed, then that indiciates ALPN failed, send failure.
+                // We have this workaround, as openssl supports terminating handshake only from version 1.1.0,
+                // whereas ALPN is supported from version 1.0.2.
+                if (done && sslAuthenticationOptions.IsServer &&
+                    sslAuthenticationOptions.ApplicationProtocols != null && ((SafeDeleteSslContext)context).SslContext.AlpnHandle.IsAllocated &&
+                    ((SafeDeleteSslContext)context).SslContext.AlpnHandle.Target == null)
+                {
+                    throw Interop.OpenSsl.CreateSslException(SR.net_alpn_failed);
```

Comment:
Do call sites expect exceptions here?  Seems like the rest of the function returns SecurityStatusPal values, which may wrap exceptions.

Reply:
yeah, the thrown exception here will be caught in the catch block later, and an internal error with this exception is returned.

Reply:
But aren't there call sites that expect exceptions won't be thrown from these Encrypt/Decrypt methods?  I think we should return the error wrapped in the appropriate wrapper.

Reply:
The HandhshakeInternal method itself catches the exception and throws internal error [here] (https://github.com/Priya91/corefx-1/blob/3fd8c3d2ab1a0c623985c54a2c8e89d3b3bcfbfe/src/System.Net.Security/src/System/Net/Security/SslStreamPal.Unix.cs#L180). I'm not sure what Encrypt/Decrypt you are referring to, this codepath will not be hit for encrypt/decrypt